WordPress/gutenberg
# Patterns: disable editing of synced patterns in post editor unless children have connected content set

A draft alternative to https://github.com/WordPress/gutenberg/pull/56574

## What?
This version edits the source pattern in place instead of redirecting to the editor

Locks pattern editing in the post editor to only child blocks that have had content connections set.

**N.B.** This PR changes the paradigm for editing synced patterns in the editor and moves the editing of a pattern's content that causes global changes to the site editor or the wp-admin pattern editing interfaces (details of why below).

## Why?
As part of the work on allowing [partial editing of synced patterns](https://github.com/WordPress/gutenberg/issues/53705) we want to disable full editing of the synced pattern entities by default in the post editor.

This will make it easier to indicate to users when they are editing just a section of the pattern for that instance of the pattern only, versus editing the global source pattern instance.

This would also fix https://github.com/WordPress/gutenberg/issues/54442 as the edit button only shows if the user has permission to update the pattern. It also fixes #32353

## How?
Sets each of the patterns innerBlocks edit mode to `disabled` using `setBlockEditingMode` unless the block connected attributes are set in which case the block is set to `contentOnly`. 

**Important note** - limiting editing to `contentOnly` in the post editor is 🤞 just the starting point. We are keeping it restricted to contentOnly to keep things simple while we set up the initial APIs for implementing partial syncing. The hope is that we can later extend it out to things like innerBlocks, limited block style attributes, etc. when/if the complexities around doing so are resolved.

## Testing Instructions
- Add a synced pattern 
- Insert the pattern into the block editor and check that the child blocks can't be edited, and don't appear in the list view
- Check that the `Edit parent pattern` button in the block toolbar unlocks the innerBlocks for full editing, and `Stop editing parent pattern` locks the innerBlocks again
- Check that the `Edit parent pattern` button does not display if the user does not have edit permission for the given pattern
- Add a new synced pattern using the following block markup, add the block to a post and make sure the Heading can't be edited but the paragraphs can be edited as `contentOnly`. Also check that the `Content` panel appears in right inspector panel when block or nested content selected.

```

<h2 class="wp-block-heading">This heading must stay the same</h2>



<p>You can change this</p>



<p>And you really should change this</p>

```

## Screenshots or screencast 


https://github.com/WordPress/gutenberg/assets/3629020/ad0e10fd-9c89-486f-815d-b22e9f34a22e
